### DISCUSSION TITLE
Passing Managed Android OpenGL Context to Native Code

### DISCUSSION THREAD

**Discussion by djrecipe** [AvaloniaUI/Avalonia#11402] [Q&A] [answered]
@jp2masa provided some excellent help on how to implement `NativeControlHost` native platforms (see https://github.com/AvaloniaUI/Avalonia/discussions/11069)

Now I am seeking more help on the details of **how to draw directly onto an Avalonia control using native code on Android**. In summary, I want to pass the managed user control context to native code and then use native OpenGL calls to draw directly onto the user control. Here are my questions:

- _Native Code_ Do I need to create a new context `eglCreateContext` with a shared context with the Avalonia user control, or use `eglMakeCurrent` and draw directly to the existing user control context?
- _Managed Code_ Given the code snippet in `EmbedSample.Android.cs`: `(parent as AndroidViewControlHandle)?.View.Context`, which property(s) of this is appropriate to pass to native code as a low-level handle?

For some context (heh, not of the OpenGl variety though!), executing native OpenGL calls on a Avalonia control _in Windows_ looks something like this:

```
protected override IPlatformHandle CreateNativeControlCore(IPlatformHandle parent)
{
    // get appropriate native handle
    var handle = base.CreateNativeControlCore(parent);
    this.Handle = handle.Handle;
    // create a window to host opengl calls, which will be a child window of this control
    IntPtr result = this.Client.CreateNewWindow(handle.Handle);
    // initialize open gl using the child window to draw on
    this.Client.InitializeOpengl(result);
    return handle;
}
```

It's difficult to find the exact details of how to map OpenGL contexts between .NET and native code on Android, so I appreciate any guidance ya'll are able to offer here.

Cheers,
Justin Pierce

**Comment by djrecipe** [comment]
The native line I seem to die on is `eglCreateWindowSurface(disp, config, (NativeWindowType)hwnd, NULL);` with error:

> [libc] Fatal signal 11 (SIGSEGV), code 1 (SEGV_MAPERR), fault addr 0x4696 in tid 25939 (PlatformDemoApp), pid 25939 (PlatformDemoApp)

For hwnd I'm passing `IPlatformHandle.Handle` from managed code

**Reply by kekekeks** [reply, marked as answer]
If you are planning to draw into a separate android view using `eglCreateWindowSurface` (like on Windows), you need to create a `SurfaceView` and extract `ANativeWindow`.

**Reply by djrecipe** [reply]
@kekekeks Thanks for the response. I think the pieces are starting to fit together in my brain. 

Can you point me to any resources on creating a `SurfaceView` within Avalonia? I've tried extracting `ANativeWindow` from `(parent as AndroidViewControlHandle).View.RootSurfaceControl.Handle` but this was just a guess and it didn't work.

Here is some pseudo-code of my current understanding:
```
// create surface view (native? managed?)
surface_view = ... ???
// extract native window
native_window = ANativeWindow_fromSurface(JNIEnv.Handle, surface_view.Handle);
// opengl init
egl_surface = eglCreateWindowSurface(disp, config, native_window, NULL);
eglMakeCurrent(disp, egl_surface, egl_surface, context);
```

As you can tell, I'm somewhat new to the Android ecosystem, but I think I've almost got it nailed down 🤞🏼

**Reply by djrecipe** [reply]
It looks like actually I might be able to create a new surface view via `(parent as AndroidViewControlHandle).View.Context`. Again, just a guess - further investigation is needed.

**Reply by kekekeks** [reply]
You can create the SurfaceView using Xamarin.Android API bindings. Context can be extracted from the parent view handle, yes.

**Reply by djrecipe** [reply]
Wonderful, thanks mate, let me give it a shot and then mark you as answered 🥇

**Reply by djrecipe** [reply]
Progress has been made! Here is my current code:

_Managed_ (in `NativeControlHost.CreateNativeControlCore`)
```
SurfaceView view = new SurfaceView((parent as AndroidViewControlHandle).View.Context);
InvokeNativeCode(JNIEnv.Handle, view.Holder.Surface.Handle);
```

_Native_
```
result = ANativeWindow_fromSurface((JNIEnv*)javaHandle, (jobject)jobjectHandle);
assert(result != nullptr);
```

ANativeWindow_fromSurface no longer throws an exception! However, it returns null.

Is there an event I should be waiting for? (I see mentions of `onSurfaceCreated`, etc.) Or perhaps I should check-out `GLSurfaceView`?

We're probably almost beyond the scope you're willing to help me with, but I'd really love to use Avalonia+.NET as a vehicle to deliver native OpenGL code to different platforms.

I'll keep playing around with it, probably investigating `GLSurfaceView` and perhaps it may be able to simply invoke native code within its various callbacks, without having to set-up an OpenGL context or "surface".

If anyone knows how to do this, please let me know!

**Reply by djrecipe** [reply]
Thankfully finally making tons of progress. The trick seems to lie in `GLSurfaceView.IRenderer`

Will post more within a week or so, so others may have an easier time getting started with this :)

**Reply by djrecipe** [reply]
@kekekeks Using the renderer of `GLSurfaceView` I can now create new window in native code, or just use `eglGetCurrentContext` which is what I'm currently using.

However, I only see a black box even when using `glClearColor` to a different color. As previously mentioned, I will continue to investigate...

My understanding of Avalonia is that it uses OpenGL behind the scenes anyway, so I look forward to working with you all to be able to implement native OpenGL content in Avalonia controls.

**Comment by djrecipe** [comment]
**Everything is now working as I want!!!** Here is the breakdown of what was required in order to meet my requirements:

**Implement `NativeControlHost`**
```
protected override IPlatformHandle CreateNativeControlCore(IPlatformHandle parent)
{
    var android_context = (parent as AndroidViewControlHandle).View.Context;
    GLSurfaceView view = new GLSurfaceView(android_context);
    view.SetRenderer(new RenderProxy(this.Client, view));
    return new AndroidViewControlHandle(view);
}
```

**Implement `GLSurfaceView.IRenderer`**
_Note: NO OpenGL initialization is required, NO calls to `eglCreateWindowSurface` etc are required_
```
public void OnSurfaceCreated()
{
   // set various opengl flags, etc - once again , you dont need to actually initialize opengl or a surface
   PInvokeNativeMethodA();
}
public void OnSurfaceChanged()
{
   // update the perspective
   PInvokeNativeMethodB();
}
public void OnDrawFrame()
{
   // your rendering work is done here
   PInvokeNativeMethodC();
}
```
